draft-barnes-mls-protocol.md

```diff
@@ -1435,44 +1434,219 @@ all arrive at the following state:
+
+The traffic keying material is generated from an input traffic secret value using:
+
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app nonce", [sender], nonce_length)
+
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app key", [sender], key_length)
+~~~~
+
+In this figure, [sender] represents the four-byte value representing the
+participant index in the tree used for the group key establishment mechanism.
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitely inherit this ownership.
+
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
+
+
+## Message Encryption and Decryption
+
+The Group participants should use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        uint32 length;
+        opaque content[length];
+        opaque signature[signature_length];
+        uint8 zeros[length_of_padding];
+    } MLSPlaintext;
+
+    struct {
+        opaque group<0..2^32-1>;
+        uint32 epoch_secret;
+        uint32 generation;
+        uint32 sender;
+        uint32 length;
+        opaque encrypted_content[MLSCiphertext.length];
+    } MLSCiphertext;
+~~~~~
 
-Alternatively, the hash of the previous message can also be included as an additional
-field rather than change the encryption key. This allows for a more flexible approach,
-because the receiving party can choose to ignore it (if the value is not known, or if
-transcript security is not required).
+The Group identifier and epoch allows a device to know which Group secret
+should be used and from which Epoch secret to start computing other secrets
+and keys. The participant identifier is used to derive the participant
+Application secret chain from the initial shared Application secret.
+The application message-counter field is used to determine which Application
+secret should be used from the chain to compute the correct AEAD keys
+before performing decryption.
+
+The signature field, allows to privide strong authentication of the
+plaintext and avoid Group participants to impersonate other participants.
+Note, that this non-repudiability property does not necessarily contradict
+deniability (see {{authentication}}).
+
+[[ OPEN ISSUE: Should the padding be required for Handshake messages ?
+Can an adversary get more that the position of a participant in the tree
+without padding ? Should the base ciphertext block length be negotiated or
+is is reasonnable to allow leaking an interval for the length of the plaintext
+by allowing to send a variable number of ciphertext blocks ? ]]
+
+Padding of Application messages SHOULD be enforced to provide some resistance
+against traffic analysis techniques over encrypted traffic.
```

Comment:
What does enforced mean here? That the receiver or the sender?

Reply:
I rephrased to something like "Application messages should be padded to provide some resistance to traffic analysis"